### PR TITLE
Lock the buildsystem while logging in

### PR DESCRIPTION
**Pull request by puiterwijk**
Signed-off-by: Patrick Uiterwijk <puiterwijk@redhat.com>

### CODE REVIEW

**Comment by bowlofeggs** [comment]
@puiterwijk Why do we want to do this? Also, this will need tests.

**Comment by puiterwijk** [comment]
Because having two sessions that login at the exact same moment confuses kerberos.

**Comment by bowlofeggs** [comment]
@puiterwijk Bodhi does have multiple processes, though I'm not sure which of them use the buildsys module. Also, I'm not sure how we have httpd configured in production (does it use processes and threads… just threads… just processes?) Since this commit only uses thread locking, should we be concerned about the processes causing this confusion?

**Comment by puiterwijk** [comment]
The problematic part is the mash stuff. It uses multiple threads, and LOTS of calls to the buildsystem, so LOTS of chances to hit the race condition.
HTTPD side is not a problem.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
This needs tests.

**Comment by jeremycline** [comment]
We obviously need to get this into Bodhi for now, but it seems like upstream Koji would benefit from making the login function thread-safe. Do you know if there's an issue or PR for that? I don't see either, but that doesn't mean it doesn't exist, of course.

**Comment by bowlofeggs** [comment]
On Wed, 2017-01-18 at 10:40 -0800, Jeremy Cline wrote:
> We obviously need to get this into Bodhi for now, but it seems like
> upstream Koji would benefit from making the login function thread-
> safe. Do you know if there's an issue or PR for that? I don't see
> either, but that doesn't mean it doesn't exist, of course.

I'm not sure, but I'm a huge +1 to the suggestion of sending this to
Koji. We can keep this code for now until they have it.

File: bodhi/server/buildsys.py [hunk 1]
```diff
@@ -314,18 +315,20 @@ def get_krb_conf(config):
...
+        return _buildsystem()
 
 
 def setup_buildsystem(settings):
-    global _buildsystem, _koji_hub
+    global _buildsystem, _koji_hub, _buildsystem_login_lock
     if _buildsystem:
         return
 
+    _buildsystem_login_lock = Lock()
```
Comment (jeremycline) [inline bodhi/server/buildsys.py:331]: Out of curiosity, why not just initialize this at the module level rather than involving ``global`` and a function?

Comment (puiterwijk) [reply bodhi/server/buildsys.py:331]: To be consistent with the rest of the code.

Comment (jeremycline) [reply bodhi/server/buildsys.py:331]: Okay, cool. I'm inclined to change it to initialize at the module level since that's less surprising to me.

**Comment by jeremycline** [comment]
I added unit tests on top of this commit in #1230 and had my question answered here so I'm going to go ahead and close this.

**Pull request closed without merging**